sdks/java/io/hadoop-format/src/main/java/org/apache/beam/sdk/io/hadoop/format/HadoopFormatIO.java

```diff
@@ -197,99 +425,764 @@ public void validate(PipelineOptions pipelineOptions) {}
+
+    /**
+     * Creates or reuses existing {@link TaskID} for given record.
+     *
+     * <p>The {@link TaskID} creation is based on the calculation hash function of {@code KeyT} of
+     * the pair via {@link Partitioner} (stored in configuration)
+     *
+     * @param kv keyvalue pair which should be written
+     * @param config hadoop configuration
+     * @return TaskID assigned to given record
+     */
+    private TaskID createTaskIDForKV(KV<KeyT, ValueT> kv, Configuration config) {
+      int taskContextKey =
+          getPartitioner(config).getPartition(kv.getKey(), kv.getValue(), getReducersCount(config));
+
+      return partitionToTaskContext.computeIfAbsent(
+          taskContextKey, (key) -> HadoopUtils.createTaskID(getJobId(config), key));
+    }
+
+    private JobID getJobId(Configuration config) {
+      if (jobId == null) {
+        jobId = HadoopUtils.getJobId(config);
+      }
+      return jobId;
+    }
+
+    private int getReducersCount(Configuration config) {
+      if (reducersCount == null) {
+        reducersCount = HadoopUtils.getReducersCount(config);
+      }
+      return reducersCount;
+    }
+
+    private Partitioner<KeyT, ValueT> getPartitioner(Configuration config) {
+      if (partitioner == null) {
+        partitioner = HadoopUtils.getPartitioner(config);
+      }
+      return partitioner;
+    }
+  }
+
+  /**
+   * Writes all {@link KV}s pair for given {@link TaskID} (Task Id determines partition of writing).
+   *
+   * <p>For every {@link TaskID} are executed following steps:
+   *
+   * <ul>
+   *   <li>Creation of {@link TaskContext} on start of bundle
+   *   <li>Writing every single {@link KV} pair via {@link RecordWriter}.
+   *   <li>Committing of task on bundle finish
+   * </ul>
+   *
+   * @param <KeyT> Type of key
+   * @param <ValueT> Type of value
+   */
+  private static class WriteFn<KeyT, ValueT> extends DoFn<KV<Integer, KV<KeyT, ValueT>>, Integer> {
+
+    private TaskContext<KeyT, ValueT> bundleTaskContext;
+    private Set<BoundedWindow> boundedWindowSet;
+
+    private PCollectionView<Configuration> configView;
+
+    WriteFn(PCollectionView<Configuration> configView) {
+      this.configView = configView;
+    }
+
+    /** Deletes cached map from previous bundle. */
+    @StartBundle
+    public void startBundle() {
+      bundleTaskContext = null;
+      boundedWindowSet = new HashSet<>();
     }
 
     @ProcessElement
-    public void processElement(ProcessContext c)
-        throws ExecutionException, InterruptedException, IOException {
-      recordWriter.write(c.element().getKey(), c.element().getValue());
+    public void processElement(
+        @Element KV<Integer, KV<KeyT, ValueT>> element, ProcessContext c, BoundedWindow b)
+        throws IOException, InterruptedException {
+
+      Configuration conf = c.sideInput(configView);
+      Integer taskID = element.getKey();
+
+      if (bundleTaskContext == null) {
+        bundleTaskContext = setupTask(taskID, conf);
```

Comment:
IMPORTANT: this seems weird, I think that we can receive multiple windowed taskids in the same bundle, am I missing something?

Reply:
This will be probably more complicated..
I have found this article: https://beam.apache.org/documentation/execution-model/ which describes how are the data distributed across bundles. I will probably have to move task setup and commit from WriteFn to separate ParDos. 
I will look on that more closely tommorow.

Reply:
I still doubt if the WriteFn is written correctly.
There are following points which should be changed by my opinion:
- Setup of the tasks should be moved into taskIds creation
 - Hash branch can do this for every partition after job setup
 - NonHash branch can do it after taskId lock acquisition
- Commit of tasks should be done differently for Hash and non Hash branch:
 - Hash branch should commit task after all writes are done - for example before commitJob
 - NonHash branch should commit task after bundle processing
- My final question is releated to [dependent parallellism](https://beam.apache.org/documentation/execution-model/#dependent-parallellism) [failures hadnling](https://beam.apache.org/documentation/execution-model/#failures-within-one-transform) in relation to WriteFn. If I understand to the document correctly than we cannot expect that KV pairs of one taksId will be processed in one bundle (at least for Hashing branch). 

I am not sure about my thoughts, so I will need some advice probably...